colexec: remove redundant deep copies in distinct template

yuzefovich opened the pull request:
**colexec: run distinct benchmarks on Bytes columns**

This commit makes it so that we now run all distinct benchmarks on Bytes
columns in addition to Ints. In order to not double the running time, we
now hardcode running only for 2 distinct columns (4 column option is now
removed since it wasn't giving us much additional info anyway).

Release note: None

**colexec: remove redundant deep copies in distinct template**

Previously, we would perform a deep copy of the current value to be
stored as "last value" when running ordered distinct (also reused by the
sorter). The deep copy is only necessary after we have processed the
batch in ordered distinct and not necessary at all for sort partitioners,
yet previously we were doing a deep copy on each value. This commit fixes
the issue which should give a nice speed up for non-trivial types.

Release note: None

yuzefovich committed:
colexec: remove redundant deep copies in distinct template

Previously, we would perform a deep copy of the current value to be
stored as "last value" when running ordered distinct (also reused by the
sorter). The deep copy is only necessary after we have processed the
batch in ordered distinct and not necessary at all for sort partitioners,
yet previously we were doing a deep copy on each value. This commit fixes
the issue which should give a nice speed up for non-trivial types.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63820)
<!-- Reviewable:end -->

yuzefovich commented:
Our microbenchmarks don't show any difference, but they are using only ints where this change doesn't matter.

jordanlewis reviewed (approved):
Great idea, this is a good point. Care to add a quick microbench for strings? I think that would be a useful one to have in the back pocket.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63820#-:-MY_jiqf3We-KlHWeIRv:bgtyq1v)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec: run distinct benchmarks on Bytes columns

This commit makes it so that we now run all distinct benchmarks on Bytes
columns in addition to Ints. In order to not double the running time, we
now hardcode running only for 2 distinct columns (4 column option is now
removed since it wasn't giving us much additional info anyway).

Release note: None

yuzefovich commented:
Added another commit with distinct benchmarks on Bytes type, and they do show an expected [speedup](https://gist.github.com/yuzefovich/3b515d11ac282bdefd0fa9f5937087bf).

yuzefovich commented:
TFTR!

bors r+

jordanlewis commented:
Very nice speedup!
